### PR TITLE
Tune heuristic defaults and behavior for improved stability

### PR DESCRIPTION
**Pull request by earthling-amzn**
Also, some minor changes to logging.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Reviewers
 * [Kelvin Nilsen](https://openjdk.org/census#kdnilsen) (@kdnilsen - Committer) ⚠️ Review applies to [7e7dfa41](https://git.openjdk.org/shenandoah/pull/209/files/7e7dfa413534219cc3660189d643634cfcb4aaaf)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/shenandoah pull/209/head:pull/209` \
`$ git checkout pull/209`

Update a local copy of the PR: \
`$ git checkout pull/209` \
`$ git pull https://git.openjdk.org/shenandoah pull/209/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 209`

View PR using the GUI difftool: \
`$ git pr show -t 209`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/shenandoah/pull/209.diff">https://git.openjdk.org/shenandoah/pull/209.diff</a>

</details>

### CODE REVIEW

**Review by kdnilsen — APPROVED** [review]
Thanks.

Ok after whitespace fixes.

File: src/hotspot/share/gc/shenandoah/shenandoah_globals.hpp [hunk 1]
```diff
@@ -147,7 +147,7 @@
           "cases. In percents of (soft) max heap size.")                    \
           range(0,100)                                                      \
                                                                             \
-  product(uintx, ShenandoahLearningSteps, 5, EXPERIMENTAL,                  \
+  product(uintx, ShenandoahLearningSteps, 10, EXPERIMENTAL,                 \
```
Comment (ysramakrishna) [inline src/hotspot/share/gc/shenandoah/shenandoah_globals.hpp:150]: Relatedly, I'd suggest just removing the "(5)" in this comment in `shenandoahGeneration.cpp`:

```
  // Changing the size of the generation will reset the times learned for the heuristic. The heuristic will need to
  // relearn collection performance metrics. This also has the effect of preventing further capacity changes from the
  // heuristics until at least ShenandoahLearningSteps(5) number of cycles has completed.
  void increase_capacity(size_t increment);
  void decrease_capacity(size_t decrement);
```

**Review by ysramakrishna — APPROVED** [review]
LGTM, modulo question re (high level) performance data to share, if any.

File: src/hotspot/share/gc/shenandoah/shenandoah_globals.hpp (hunk 1 above)
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoah_globals.hpp:150]: That whole comment is out of date, I'll remove it.

File: src/hotspot/share/gc/shenandoah/shenandoah_globals.hpp [hunk 2]
```diff
@@ -181,7 +181,7 @@
           "the heuristic is to allocation spikes. Decreasing this number "  \
           "increases the sensitivity. ")                                    \
                                                                             \
-  product(double, ShenandoahAdaptiveDecayFactor, 0.5, EXPERIMENTAL,         \
+  product(double, ShenandoahAdaptiveDecayFactor, 0.1, EXPERIMENTAL,         \
```
[the remark this answers is not in the document]
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoah_globals.hpp:184]: Yes! The maximum number of degenerated cycles observed on 10 runs of `specjbb2015` went from 48 to 9. Same metric for `extremem` went from 111 to 86. The effect is less dramatic for satb mode. There are also notable improvements on Dacapo's `jython` and `biojava`:
```
+11.58% hyperalloc_a2048_o1536/context_switch_count p=0.00537
  Control:   2515.458   (+/-249.10  )         90
  Test:      2806.800   (+/-145.41  )          5

+10.97% hyperalloc_a3072_o1536/context_switch_count p=0.00002
  Control:   3780.958   (+/-377.60  )         90
  Test:      4195.600   (+/-103.85  )          5

+10.14% extremem/product_replacement_p50 p=0.00147
  Control:      1.793ms (+/-125.44us)         85
  Test:         1.975ms (+/-184.30us)          5

+5.33% hyperalloc_a2048_o1536/cpu_user p=0.00004
  Control:    287.015s  (+/-  7.43s )         90
  Test:       302.300s  (+/-  7.52s )          5

-162.25% jython/rss_max p=0.00000
  Control: 6410130.783   (+/-65440.95  )         85
  Test:    2444264.800   (+/-73349.32  )          5

-153.73% jython/minor_page_fault_count p=0.00000
  Control: 1682775.464   (+/-18501.36  )         85
  Test:    663203.400   (+/-22693.20  )          5

-146.98% biojava/minor_page_fault_count p=0.00000
  Control: 1566706.957   (+/-1318.36  )         85
  Test:    634354.600   (+/-17082.50  )          5

-77.56% jython/cpu_system p=0.00000
  Control:      3.864s  (+/-  0.15s )         85
  Test:         2.176s  (+/-  0.12s )          5

-38.79% extremem-phased/calculate_addresses p=0.00000
  Control:    433.731ms (+/- 58.95ms)        795
  Test:       312.516ms (+/- 32.93ms)         12

-18.63% extremem-phased/minor_page_fault_count p=0.00000
  Control: 4272509.681   (+/-243509.95  )         85
  Test:    3601418.400   (+/-70657.90  )          5
```

**Pull request closed without merging**